apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -111,33 +118,44 @@ internal_resolve(Protocol, DNSMessage) ->
     Response = erldns_handler:do_handle(DNSMessage, ?LOCALHOST),
     encode_message(Protocol, Response).
 
--spec(resolve(protocol(), [upstream()], binary(), reply_fun()) -> ok).
-resolve(Protocol, Upstreams, Request, Fun) ->
+-spec(external_resolve(protocol(), [upstream()], binary(), reply_fun(), binary(), zone())-> ok).
+external_resolve(Protocol, Upstreams, Request, Fun, FailMsg, Zone) ->
     Workers =
         lists:map(fun (Upstream) ->
             Pid = start_worker(Protocol, Upstream, Request),
             MonRef = monitor(process, Pid),
             {Upstream, Pid, MonRef}
         end, Upstreams),
-    resolve_loop(Workers, Fun).
+    resolve_loop(Workers, Fun, FailMsg, Zone).
+
+-spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun(), binary(), zone()) -> ok).
+resolve_loop(Workers, Fun, FailMsg, Zone) ->
+    resolve_loop(Workers, Fun, FailMsg, Zone, unresolved).
 
--spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun()) -> ok).
-resolve_loop([], _Fun) ->
+-spec(resolve_loop([{upstream(), pid(), reference()}], reply_fun(), binary(), zone(), forward_status()) -> ok).
+resolve_loop([], _Fun, _Zone, _FailMsg, resolved) ->
     ok;
-resolve_loop(Workers, Fun) ->
+resolve_loop([], Fun, Zone, FailMsg, Status) ->
+    ok = reply(Fun, FailMsg),
+    prometheus_counter:inc(dns_forwarder_failures_total, [Zone, Status]);
+resolve_loop(Workers, Fun, Zone, FailMsg, Status) ->
     receive
         {reply, Pid, Response} ->
             {value, {Upstream, Pid, MonRef}, Workers0} =
                 lists:keytake(Pid, 2, Workers),
-            reply(Fun, Response),
+            ok = reply(Fun, Response),
             erlang:demonitor(MonRef, [flush]),
             report_status([?MODULE, Upstream, successes]),
-            resolve_loop(Workers0, fun ok_reply_fun/1);
+            resolve_loop(Workers0, fun ok_reply_fun/1, FailMsg, Zone, resolved);
         {'DOWN', MonRef, process, Pid, _Reason} ->
             {value, {Upstream, Pid, MonRef}, Workers0} =
                 lists:keytake(Pid, 2, Workers),
             report_status([?MODULE, Upstream, failures]),
-            resolve_loop(Workers0, Fun)
+            Status0 = case Status of
+                resolved -> resolved;
+                _ -> timeout
+            end,
+            resolve_loop(Workers0, Fun, FailMsg, Zone, Status0)
     after 2 * ?TIMEOUT ->
```

Comment:
I guess we discussed it, but to triple check. My understanding is that the `Pid ! {timeout, self()}` will be capture by the respective workers (line 212 for udp and line 254 tcp), which will prompt those process to exit normally, which will trigger line 150, resulting on responding with the `failMsg` and emitting a `failure` metric.